geoprocessing/src/modules/tile/tile.service.ts

```diff
@@ -0,0 +1,313 @@
+ * @description Input interface for the tile query builder
+ */
+export interface ITileQuery {
+  x: number;
+  y: number;
+  z: number;
+  table: string;
+  geometry: string;
+  extent: number;
+  buffer: number;
+}
+
+/**
+ * @description The tile query builder callback definition
+ */
+export type GetTileQuery = (input: ITileQuery) => string;
+
+/**
+ * @description Input interface for the base query builder
+ */
+export interface IBaseQueryInput {
+  x: number;
+  y: number;
+  z: number;
+  table: string;
+  geometry: string;
+  maxZoomLevel: number;
+  buffer: number;
+  extent: number;
+  customQuery: string;
+  attributes: string;
+}
+
+/**
+ * @description The base query builder callback definition
+ */
+export type GetBaseQuery = (input: IBaseQueryInput) => string;
+
+/**
+ * @description This function define tile query input for the tile requested
+ */
+export interface ITileQueryInput extends TileRequest {
+  z: number;
+  x: number;
+  y: number;
+  maxZoomLevel: number;
+  table: string;
+  geometry: string;
+  extent: number;
+  buffer: number;
+  customQuery: string;
+  attributes: string;
+  getBaseQuery?: GetBaseQuery;
+}
+
+@Injectable()
+export class TileService {
+  /**
+   * @todo add constructor
+   * @todo move generation of specific query for each point to the api. Generate this query with the query builder
+   * @todo fix geometry issue with gid_0 = 'ATA'. Once is fixed, remove this condition from the query.
+   * @description The default base query builder
+   */
+  private readonly logger: Logger = new Logger(TileService.name);
+  defaultGetBaseQuery: GetBaseQuery = ({
+    x,
+    y,
+    z,
+    table,
+    geometry,
+    extent,
+    buffer,
+    customQuery,
+    attributes,
+  }: IBaseQueryInput) => `
+  SELECT
+    ${attributes},
+    ST_AsMVTGeom(
+      -- Geometry from table
+      ST_Transform(ST_RemoveRepeatedPoints(${geometry}, 0.1), 3857),
+      -- MVT tile boundary
+      ST_TileEnvelope(${z}, ${x}, ${y}),
+      -- Extent
+      ${extent},
+      -- Buffer
+      ${buffer},
+      -- Clip geom
+      true
+    ) AS mvt_geom
+  FROM ${table}
+  WHERE
+    ST_Intersects(ST_Transform(ST_TileEnvelope(${z}, ${x}, ${y}), 4326), ${geometry})
+    and ${customQuery}
+    and not gid_0 ='ATA'
+  `;
+
+  /**
+   * @description Creates the query for tile
+   */
+  createQueryForTile({
+    z,
+    x,
+    y,
+    maxZoomLevel,
+    table,
+    geometry,
+    extent,
+    buffer,
+    customQuery,
+    attributes,
+    getBaseQuery = this.defaultGetBaseQuery,
+  }: ITileQueryInput): string {
+    const queryParts: string[] = [];
+    queryParts.push(
+      `WITH tile AS (${getBaseQuery({
+        x,
+        y,
+        z,
+        table,
+        geometry,
+        maxZoomLevel,
+        buffer,
+        extent,
+        customQuery,
+        attributes,
+      })})`,
+    );
+
+    const sql = `${queryParts.join(
+      ',\n',
+    )}\nSELECT ST_AsMVT(tile, 'layer0',  ${extent}, 'mvt_geom') AS mvt FROM tile`;
```

Comment:
is there an expectatino for `queryParts` to grow beyond a one-element array? or maybe this is left over from an initial implementation? As it stands, I think this could be moved up one level in the call stack, also avoiding the `.push()`, and `.join()` parts, since in fact this will always be (in the current implementation) a one-element array.

Reply:
As highlighted above, I've merged all the query functions so we are avoiding these steps (push and join for example)